REGRESSION (294844@main): Choppy scrolling on kottke.org

smfr opened the pull request:
#### b42f9106d95496499bb100f76585a40fc666d7c0
<pre>
REGRESSION (294844@main): Choppy scrolling on kottke.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=295434">https://bugs.webkit.org/show_bug.cgi?id=295434</a>
<a href="https://rdar.apple.com/152254822">rdar://152254822</a>

Reviewed by Abrar Rahman Protyasha.

kottke.org has some repeating gradients with `background-attachment:fixed`,
which means they paint on every scroll. The color stop colorspace conversion
added in 294844@main runs on every stop, so we end up doing hundreds of
redundant conversions per paint.

Fix by adding a 32-entry cache of converted color components, which is
cleared if the destination colorspace changes. It&apos;s locked since this code
has to be threadsafe for offscreen canvas.

I considered using a TinyLRUCache, but it&apos;s entry creation callback doesn&apos;t
have a way to provide context.

* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::getResolvedColorComponentsInColorSpace):
(WebCore::GradientRendererCG::makeGradient const):

Canonical link: <a href="https://commits.webkit.org/297029@main">https://commits.webkit.org/297029@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4146bc6dff659e8b397d8fe472238db040f709dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110207 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116229 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60455 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38453 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83792 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113155 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24361 "Found 1 new test failure: fast/filter-image/clipped-filter.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99243 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23725 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17377 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60025 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93738 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119020 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37247 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27616 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92761 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37619 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95513 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92585 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23613 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37583 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15316 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33136 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37141 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42612 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36803 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38512 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
```diff
@@ -76,6 +78,38 @@ GradientRendererCG::Strategy GradientRendererCG::pickStrategy(ColorInterpolation
+        return color.toResolvedColorComponentsInColorSpace(destinationColorSpace);
+    }).iterator->value;
+
+    const size_t maxCacheCount = 32;
```
`static constexpr`

aprotyas commented on Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
```diff
@@ -121,7 +155,7 @@ GradientRendererCG::Strategy GradientRendererCG::makeGradient(ColorInterpolation
             for (const auto& stop : stops) {
                 if (destinationColorSpace) {
-                    auto [r, g, b, a] = stop.color.toResolvedColorComponentsInColorSpace(*destinationColorSpace);
+                    auto [r, g, b, a ] = getResolvedColorComponentsInColorSpace(stop.color, *destinationColorSpace);
```
Nit: don't need that space
```suggestion
                    auto [r, g, b, a] = getResolvedColorComponentsInColorSpace(stop.color, *destinationColorSpace);
```

smfr committed:
REGRESSION (294844@main): Choppy scrolling on kottke.org
https://bugs.webkit.org/show_bug.cgi?id=295434
rdar://152254822

Reviewed by Abrar Rahman Protyasha.

kottke.org has some repeating gradients with `background-attachment:fixed`,
which means they paint on every scroll. The color stop colorspace conversion
added in 294844@main runs on every stop, so we end up doing hundreds of
redundant conversions per paint.

Fix by adding a 32-entry cache of converted color components, which is
cleared if the destination colorspace changes. It's locked since this code
has to be threadsafe for offscreen canvas.

I considered using a TinyLRUCache, but it's entry creation callback doesn't
have a way to provide context.

* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::getResolvedColorComponentsInColorSpace):
(WebCore::GradientRendererCG::makeGradient const):

Canonical link: https://commits.webkit.org/297029@main
